### PR TITLE
POC multiple active gateways

### PR DESCRIPTION
**Pull request by astoycos**
Rebase the feature branch and Implement the Core submariner bits for Multiple Active Gateways POC.  For design information, and further details please see the accompanying EP. 

MAG Enhancement Proposal -> https://github.com/submariner-io/enhancements/pull/69

### CODE REVIEW

**Comment by astoycos** [comment]
Still have a bit of work before unit tests will pass, and I will need to debug the failing e2e tests.

**Comment by astoycos** [comment]
Expecting Globalnet tests to fail here since complete functionality has not yet been implemented within the POC

**Comment by astoycos** [comment]
So in followup to the above comment, the following failing tests 

```
End to End Full / E2E (operator, globalnet, libreswan) (pull_request) 

End to End Full / E2E (operator, globalnet, wireguard) (pull_request) 

End to End Full / E2E (operator, globalnet, vxlan) (pull_request)

End to End Full / External Network E2E (external-net, globalnet) (pull_request) 

``` 

are expected since the full Globalnet controller (headless service functionality) has not been implemented in order for MAG basic cross cluster service functionality to work 

Also PR contains POC code so unit test coverage has not been 100% complete, however they have been fixed enough to pass in CI

**Comment by astoycos** [comment]
additionally 

```
Linting / Commit Message(s) (pull_request) 

Linting / Markdown Links (modified files) (pull_request) 
```
Are permafailing

File: pkg/cableengine/syncer/syncer.go [hunk 1]
```diff
@@ -167,6 +167,7 @@ func (gs *GatewaySyncer) cleanupStaleGatewayEntries(localGatewayName string) err
 
 		if stale {
 			klog.Warningf("stale gateway: %s, didn't report for %s", gw.Name, GatewayStaleTimeout)
+
```
Comment (astoycos) [inline pkg/cableengine/syncer/syncer.go:170]: useless space

File: pkg/cable/vxlan/vxlan.go [hunk 2]
```diff
@@ -257,7 +257,8 @@ func (v *vxlan) getVxlanVtepIPAddress(ipAddr string) (net.IP, error) {
 	return vxlanIP, nil
 }
 
-func (v *vxlan) ConnectToEndpoint(endpointInfo *natdiscovery.NATEndpointInfo) (string, error) {
+func (v *vxlan) ConnectToEndpoint(endpointInfo *natdiscovery.NATEndpointInfo) (string, error) { // nolint // Supposed to be complicated
```
Comment (astoycos) [inline pkg/cable/vxlan/vxlan.go:260]: I should move all of these cabledrive changes to an earlier commit, to better segment from globalnet changes.

File: pkg/util/util.go [hunk 3]
```diff
@@ -135,6 +135,16 @@ func CompareEndpointSpec(left, right *subv1.EndpointSpec) bool {
 		left.Backend == right.Backend && equality.Semantic.DeepEqual(left.BackendConfig, right.BackendConfig)
 }
 
+func CompareSlices(left, right []string) bool {
```
Comment (astoycos) [inline pkg/util/util.go]: Remove, this isn't used

File: pkg/cableengine/cableengine.go [hunk 4]
```diff
@@ -131,7 +131,7 @@ func (i *engine) SetupNATDiscovery(natDiscovery natdiscovery.Interface) {
 	}()
```
Comment (astoycos) [inline pkg/cableengine/cableengine.go:131]: This should also be moved to an earlier commit

**Comment by astoycos** [comment]
@tpantelis Since you helped before :) Can you please make 

```
End to End Full / E2E (operator, globalnet, libreswan) (pull_request) 

End to End Full / E2E (operator, globalnet, wireguard) (pull_request) 

End to End Full / E2E (operator, globalnet, vxlan) (pull_request)

End to End Full / External Network E2E (external-net, globalnet) (pull_request) 

Linting / Commit Message(s) (pull_request) 

Linting / Markdown Links (modified files) (pull_request) 
```

Optional for the feature branch so that we can merge this in once everything is tided up.

**Review by anfredette — APPROVED** [review]
lgtm

**Comment by anfredette** [comment]
@tpantelis @sridhargaddam @skitt Per Andrews's comments above, this code represents a POC of the multiple active gateway proposal.  It does not include support for all features and, therefore, will not pass all test cases.  It's also not necessarily the actual code that would get merged into devel should we decide to proceed with this feature.  There is a discussion of a possible approach to getting the capabilities into devel in the README.

That said, we would like to merge it in as a step toward the end goal and to preserve the state of things.  To do this, we need several test cases to be changed to not required.  Are you okay with this?

Thanks,
Andre

**Review by Billy99 — APPROVED** [review]

**Comment by astoycos** [comment]
```
[20:14:40.992] [submariner]$ [cluster2] bash -c curl -Ls https://get.submariner.io/ | VERSION=feature-multi-active-gw DESTDIR=/go/bin bash
Installing subctl version vfeature-multi-active-gw
  OS detected:           linux
  Architecture detected: amd64
  Download URL:          https://github.com/submariner-io/releases/releases/download/vfeature-multi-active-gw/subctl-vfeature-multi-active-gw-linux-amd64.tar.xz

Downloading...
curl: (22) The requested URL returned error: 404
xz: (stdin): File format not recognized
tar: Child returned status 1
tar: Error is not recoverable: exiting now
[20:14:41.473] [submariner]$ [cluster2] bash -c curl -Ls https://get.submariner.io/ | VERSION=devel DESTDIR=/go/bin bash
Installing subctl version devel
  OS detected:           linux
  Architecture detected: amd64
  Download URL:          https://github.com/submariner-io/submariner-operator/releases/download/subctl-devel/subctl-devel-linux-amd64.tar.xz

Downloading...
subctl-devel-linux-amd64 has been installed as /go/bin/subctl
This provides subctl version: devel
``` 

Is blocking us now since we're getting the operator images from devel instead of the feature branch

**Pull request merged**